### PR TITLE
Fix `gef-remote`: retrieve pid of inferior process after having connected to target

### PR DESCRIPTION
**Pull request by theguy147**
### Description/Motivation/Screenshots ###

see #685

I still have to add better tests to `gef-remote` so this won't happen again but I wanted to open the PR already.

### How Has This Been Tested? ###

| Architecture |          Yes/No          | Comments                                  |
| ------------ | :----------------------: | ----------------------------------------- |
| x86-32       | :heavy_check_mark:  |  |
| x86-64       | :heavy_check_mark:  |                                           |
| ARM          | :heavy_multiplication_x: |                                           |
| AARCH64      | :heavy_multiplication_x: |                                           |
| MIPS         | :heavy_multiplication_x: |                                           |
| POWERPC      | :heavy_multiplication_x: |                                           |
| SPARC        | :heavy_multiplication_x: |                                           |
| RISC-V       | :heavy_multiplication_x: |                                           |
| `make test` | :heavy_check_mark: |  but still needs new unit tests for `gef-remote`      |

### Checklist ###

- [x] My PR was done against the `dev` branch, not `master`.
- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required.
- [ ] My change adds tests as appropriate.
- [x] I have read and agree to the **CONTRIBUTING** document.

### CODE REVIEW

File: docs/commands/gef-remote.md [hunk 1]
```diff
@@ -77,14 +78,14 @@ for `gef` will not work either, as `gef` won't be able to fetch the content of
 the remote procfs.
 
 To circumvent this and still enjoy `gef` features with QEMU-user, a simple stub
-can be artificially added, with the option `-q` option of `gef-remote`. Note
-that you need to set the architecture properly first:
+can be artificially added, with the option `--qemu-mode` option of `gef-remote`.
+Note that you need to set the architecture properly first:
```
Comment (Grazfather) [inline docs/commands/gef-remote.md:82]: For clarity let's say "set the architecture to match the target binary" or something.

File: gef.py [hunk 2]
```diff
@@ -2755,7 +2755,10 @@ def get_os():
 @lru_cache()
 def get_pid():
     """Return the PID of the debuggee process."""
-    return gdb.selected_inferior().pid if not __gef_qemu_mode__ else gdb.selected_thread().ptid[1]
+    pid = gdb.selected_inferior().pid if not __gef_qemu_mode__ else gdb.selected_thread().ptid[1]
+    if pid == 0:
+        raise RuntimeError("cannot retrieve PID for debuggee process")
```
Comment (Grazfather) [inline gef.py:2760]: I wonder whether we should use 'interior' as is standard in gdb parlance. @hugsy what do you think?

File: docs/commands/gef-remote.md (hunk 1 above)
Comment (Grazfather) [reply docs/commands/gef-remote.md:82]: "of the inferior" maybe.

Comment (theguy147) [reply docs/commands/gef-remote.md:82]: Good call. To me "target binary" sounds better than "inferior binary"

File: gef.py (hunk 2 above)
Comment (theguy147) [reply gef.py:2760]: what do you mean by using `interior`?

Comment (Grazfather) [reply gef.py:2760]: I mean `inferior` It GDB it means the target, e.g. `info inferior`: https://moss.cs.iit.edu/cs351/gdb-inferiors.html

I think `target` or `inferior`. Either way we should be consistent.

File: docs/commands/gef-remote.md (hunk 1 above)
Comment (Grazfather) [reply docs/commands/gef-remote.md]: I agree, but I'm willing to let hugsy make the call.

File: gef.py (hunk 2 above)
Comment (theguy147) [reply gef.py:2760]: ah, you mean the error message, right? i was wondering about the `pid = ...` line.

yeah, you are right, it should be more consistent.

Comment (Grazfather) [reply gef.py:2760]: Yeah, I am nit-picky about consistency and stuff :) but that's why I commented on the `raise` line.

File: gef.py [hunk 3]
```diff
@@ -2755,7 +2755,10 @@ def get_os():
 @lru_cache()
 def get_pid():
     """Return the PID of the debuggee process."""
-    return gdb.selected_inferior().pid if not __gef_qemu_mode__ else gdb.selected_thread().ptid[1]
+    pid = gdb.selected_inferior().pid if not __gef_qemu_mode__ else gdb.selected_thread().ptid[1]
+    if pid == 0:
```
Comment (Grazfather) [inline gef.py:2759]: `if not pid` is probably safer

**Comment by theguy147** [comment]
I updated the PR to also take care of the findings in #693 for `gef-remote`

File: docs/commands/gef-remote.md [hunk 4]
```diff
@@ -60,12 +61,26 @@ automatically download the target file and store in the local temporary
...
+### Handling remote libraries ###
 
-You can then reuse the downloaded file for your future debugging sessions, use it under IDA
-and such. This makes the entire remote debugging process (particularly for Android applications)
-a child's game.
+Often times you are missing a specific library the remote process is using.
+Therefore `gef-remote` can download remote libraries (and other files) if the
+remote target supports it (and the remote gdbserver has sufficient permissions).
+The `--download-lib LIBRARY` option can download a remote file specified by
+it's filepath. Furthermore `--download-everything` downloads all remote libs
```
Comment (Grazfather) [inline docs/commands/gef-remote.md:76]: it's -> its

File: docs/commands/gef-remote.md [hunk 5]
```diff
@@ -60,12 +61,26 @@ automatically download the target file and store in the local temporary
...
+it under IDA and such. This makes the entire remote debugging process
+(particularly for Android applications) a child's game.
+
+### Handling remote libraries ###
 
-You can then reuse the downloaded file for your future debugging sessions, use it under IDA
-and such. This makes the entire remote debugging process (particularly for Android applications)
-a child's game.
+Often times you are missing a specific library the remote process is using.
+Therefore `gef-remote` can download remote libraries (and other files) if the
```
Comment (Grazfather) [inline docs/commands/gef-remote.md:73]: "Therefore" doesn't make sense. Something else.

File: docs/commands/gef-remote.md [hunk 6]
```diff
@@ -60,12 +61,26 @@ automatically download the target file and store in the local temporary
...
+remote target supports it (and the remote gdbserver has sufficient permissions).
+The `--download-lib LIBRARY` option can download a remote file specified by
+it's filepath. Furthermore `--download-everything` downloads all remote libs
+found in the processes virtual memory map (`vmmap`).
```
Comment (Grazfather) [inline docs/commands/gef-remote.md:77]: process's

File: docs/commands/gef-remote.md [hunk 7]
```diff
@@ -60,12 +61,26 @@ automatically download the target file and store in the local temporary
...
 
+Another issue with libraries is that even if you have the same libraries that
+are used remotely they might have different filepaths and GDB can't
+automatically find them and thereby can't resolve their symbols. The option
+`--update-solib` adds the previously, with `--dowload-everything`, downloaded
```
Comment (Grazfather) [inline docs/commands/gef-remote.md:82]:
```suggestion
`--update-solib` adds the previously (with `--dowload-everything`) downloaded
```

File: gef.py (hunk 2 above)
Comment (Grazfather) [reply gef.py:2760]: let's use `target process` and fix in the doc string.

**Review by Grazfather — CHANGES_REQUESTED** [review]

**Pull request merged**